<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="generator" content="litedown 0.7">
<title>repo.data</title>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  print-color-adjust: exact;
  -webkit-print-color-adjust: exact;
}
body, .abstract, code, .footnotes, footer, #refs, .caption { font-size: .9em; }
li li { font-size: .95em; }
ul:has(li > input[type="checkbox"]) { list-style: none; padding-left: 1em; }
*, :before, :after { box-sizing: border-box; }
a { color: steelblue; }
pre, img { max-width: 100%; }
pre { white-space: pre-wrap; word-break: break-word; }
pre code { display: block; padding: 1em; overflow-x: auto; }
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre, th) > code, code[class], div > .caption { background: #f8f8f8; }
pre > code:is(:not([class]), .language-plain, .language-none, .plain), .box, .figure, .table { background: inherit; border: 1px solid #eee; }
pre > code {
  &.message { border-color: #9eeaf9; }
  &.warning { background: #fff3cd; border-color: #fff3cd; }
  &.error { background: #f8d7da; border-color: #f8d7da; }
}
.fenced-chunk { border-left: 1px solid #666; }
.code-fence {
  opacity: .4;
  border: 1px dashed #666;
  border-left: 2px solid;
  &:hover { opacity: inherit; }
}
.box, .figure, .table, table { margin: 1em auto; }
div > .caption { padding: 1px 1em; }
.figure { p:has(img, svg), pre:has(svg) { text-align: center; } }
.flex-col { display: flex; justify-content: space-between; }
table {
  &:only-child:not(.table > *) { margin: auto; }
  th, td { padding: 5px; font-variant-numeric: tabular-nums; }
  thead, tfoot, tr:nth-child(even) { background: whitesmoke; }
  thead th { border-bottom: 1px solid #ddd; }
  &:not(.datatable-table) {
    border-top: 1px solid #666;
    border-bottom: 1px solid #666;
  }
}
blockquote {
  color: #666;
  margin: 0;
  padding: 1px 1em;
  border-left: .5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC {
  a { text-decoration: none; }
  ul { list-style: none; padding-left: 1em; }
  & > ul { padding: 0; }
  ul ul { border-left: 1px solid lightsteelblue; }
}
.body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.main-number::after { content: "."; }
span[class^="ref-number-"] { font-weight: bold; }
.ref-number-fig::after, .ref-number-tab::after { content: ":"; }
.cross-ref-chp::before { content: "Chapter "; }
.cross-ref-sec::before { content: "Section "; }
.cross-ref-fig::before, .ref-number-fig::before { content: "Figure "; }
.cross-ref-tab::before, .ref-number-tab::before { content: "Table "; }
.cross-ref-eqn::before, .MathJax_ref:has(mjx-mtext > mjx-c + mjx-c)::before { content: "Equation "; }
.abstract, #refs {
  &::before { display: block; margin: 1em auto; font-weight: bold; }
}
.abstract::before { content: "Abstract"; text-align: center; }
#refs::before { content: "Bibliography"; font-size: 1.5em; }
.ref-paren-open::before { content: "("; }
.ref-paren-close::after { content: ")"; }
.ref-semicolon::after { content: "; "; }
.ref-and::after { content: " and "; }
.ref-et-al::after { content: " et al."; font-style: italic; }
.footnote-ref a {
  &::before { content: "["; }
  &::after { content: "]"; }
}
section.footnotes {
  margin-top: 2em;
  &::before { content: ""; display: block; max-width: 20em; }
}
.fade {
  background: repeating-linear-gradient(135deg, white, white 30px, #ddd 32px, #ddd 32px);
  opacity: 0.6;
}

@media print {
  body { max-width: 100%; }
  tr, img { break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  body:not(.pagesjs) pre:has(.line-numbers):not(:hover) { white-space: pre; }
}
</style>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@xiee/utils@1.14.14/css/prism-xcode.min.css">
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
</head>
<body>
<div class="frontmatter">
<div class="title"><h1>repo.data</h1></div>
</div>
<div class="body">
<pre><code class="language-r">options(repos = c(&quot;@CRAN@&quot; = &quot;https://CRAN.R-project.org&quot;))
library(repo.data)
</code></pre>
<p>This vignette is written for concerned package maintainers or users that want to check their package in relationship with other packages.
For general usage of the other functions check the manual or package index (a vignette might come too).</p>
<h1 id="chp:keeping-up-with-the-repositories">Keeping up with the repositories</h1>
<p>Packages required by a package might have their own dependencies with minimal versions requirements.
The maintainers or developers might wonder what is the oldest version of each recursive package their users are required to have.
This is useful for developing packages that should remain compatible with old versions of R and packages.</p>
<pre><code class="language-r">pd &lt;- package_dependencies(&quot;ggeasy&quot;)
head(pd)
#&gt;        Name Version    Type   Op Package
#&gt; 1         R   4.4.0 Depends   &gt;=    &lt;NA&gt;
#&gt; 2       cli   3.4.0 Imports   &gt;=    &lt;NA&gt;
#&gt; 3 lifecycle   1.0.3 Imports   &gt;=    &lt;NA&gt;
#&gt; 4     rlang   1.1.0 Imports   &gt;=    &lt;NA&gt;
#&gt; 5     vctrs   0.6.0 Imports   &gt;=    &lt;NA&gt;
#&gt; 6 grDevices    &lt;NA&gt;    &lt;NA&gt; &lt;NA&gt;    &lt;NA&gt;
</code></pre>
<p><code>package_dependencies()</code> identify the minimal version required for each dependency.
If no version is required by any dependencies <code>NA</code> is used.</p>
<p>We can identify packages that are required on a lower version than one of the dependencies with</p>
<pre><code class="language-r"># Discover the requirements that can be upgraded
update_dependencies(&quot;ggeasy&quot;)
#&gt;   Name Version
#&gt; 1    R   4.4.0
</code></pre>
<p>Increasing these version requirements on <code>{ggeasy}</code> won’t affect users as they already should have these versions installed as required by other dependencies.</p>
<p>We can also be interested on since when can users install a package.
There can be two possible answers:</p>
<ul>
<li>Since it was published/release on the repositories</li>
<li>If the maintainer developers are careful the requirements might be available earlier.</li>
</ul>
<p>We can use <code>package_date()</code> to get those answers:</p>
<pre><code class="language-r">package_date(&quot;ggeasy&quot;)
#&gt;                  Published             deps_available 
#&gt; &quot;2025-06-15 08:40:04 CEST&quot; &quot;2021-05-18 09:05:22 CEST&quot;
</code></pre>
<p>Why are they important?
The first one is important to know if it hasn’t been updated in a long time.
The second one helps estimate if it can be installed on old systems without updating anything else.
If the date the dependencies are available is closer to the published date, the users will need to have updated systems and dependencies.</p>
<h1 id="chp:improving-packages">Improving packages</h1>
<p>Help pages are found via alias, when a user press <code>?word</code> it searches for alias.
Checking for existing alias might help you to find packages and reduce the confusion on the help pages.</p>
<pre><code class="language-r">alias &lt;- cran_alias(c(&quot;fect&quot;, &quot;gsynth&quot;))
#&gt; Retrieving cran_aliases, this might take a bit.
#&gt; Caching results to be faster next call in this session.
dup_alias &lt;- duplicated_alias(alias)
head(dup_alias)
#&gt;          Target Package             Source
#&gt; 1         XXinv    fect   fect-internal.Rd
#&gt; 2         XXinv  gsynth gsynth-internal.Rd
#&gt; 3      Y_demean    fect   fect-internal.Rd
#&gt; 4      Y_demean  gsynth gsynth-internal.Rd
#&gt; 5 _gsynth_XXinv    fect   fect-internal.Rd
#&gt; 6 _gsynth_XXinv  gsynth gsynth-internal.Rd
</code></pre>
<p>For example these two packages have the same alias for the internal functions but most of them point to the same file.</p>
<h2 id="sec:connecting-help-pages">Connecting help pages</h2>
<p>Often it is helpful to link help pages so that:</p>
<ul>
<li>Pages are linked to other pages</li>
<li>Pages are linked from other pages.</li>
</ul>
<pre><code class="language-r">pkg &lt;- &quot;BaseSet&quot;
head(cran_help_pages_wo_links(pkg))
#&gt; Retrieving base_aliases, this might take a bit.
#&gt; Caching results to be faster next call in this session.
#&gt; Warning in cran_alias(cran_pkgs): Omitting packages bartMachineJARs, clean, fontBitstreamVera, fontLiberation, GreedyExperimentalDesignJARs, hse, LifeInsuranceContracts, openNLPdata, positron.tutorials, RKEAjars, RMOAjars, ROI.plugin.cplex, ROI.plugin.glpk, ROI.plugin.ipop, ROI.plugin.symphony, rsparkling, RWekajars, Sejong.
#&gt; Maybe they are currently not on CRAN?
#&gt; Warning: Packages 'Hmisc', 'geneHapR', 'sfsmisc' have targets not present in a
#&gt; OS.
#&gt; Retrieving cran_rdxrefs, this might take a bit.
#&gt; Caching results to be faster next call in this session.
#&gt; Warning in cran_links(cran_pkgs): Omitting packages bartMachineJARs, clean, fontBitstreamVera, fontLiberation, GreedyExperimentalDesignJARs, hse, LifeInsuranceContracts, openNLPdata, positron.tutorials, RKEAjars, RMOAjars, ROI.plugin.cplex, ROI.plugin.glpk, ROI.plugin.ipop, ROI.plugin.symphony, rsparkling, RWekajars, Sejong.
#&gt; Maybe they are currently not on CRAN?
#&gt; Warning: Some pages point to different places according to the OS.
#&gt; Warning: Some links are distinct depending on the OS.
#&gt;   Package             Source
#&gt; 1 BaseSet BaseSet-package.Rd
#&gt; 2 BaseSet   TidySet-class.Rd
#&gt; 3 BaseSet        activate.Rd
#&gt; 4 BaseSet      add_column.Rd
#&gt; 5 BaseSet    add_elements.Rd
#&gt; 6 BaseSet    add_relation.Rd
head(cran_help_pages_not_linked(pkg))
#&gt;   Package             Source
#&gt; 1 BaseSet BaseSet-package.Rd
#&gt; 2 BaseSet   TidySet-class.Rd
#&gt; 3 BaseSet        activate.Rd
#&gt; 4 BaseSet      add_column.Rd
#&gt; 5 BaseSet    add_elements.Rd
#&gt; 6 BaseSet    add_relation.Rd
</code></pre>
<p>In addition to those help pages that are not well connected it could be that some pages are linked but link to each other without connecting with other help pages of the package or other packages.</p>
<pre><code class="language-r">cliques &lt;- cran_help_cliques(pkg)
#&gt; Warning in cran_targets_links(pkges): Omitting packages BaseSet, dplyr, methods, rlang, utils, cli, generics, glue, lifecycle, magrittr, pillar, R6, tibble, tidyselect, vctrs, utf8, pkgconfig, withr, graphics, grDevices.
#&gt; Maybe they are currently not on CRAN?
table(cliques$n) # Number of help pages connected
#&gt; &lt; table of extent 0 &gt;
</code></pre>
<p>If there is more than one length this would mean some pages not linked to the rest of the package.</p>
<p>Sometimes even if links exists they might not resolve correctly on the html version.
For example if they link to a help page of a package that is not on the strong dependency list.</p>
<pre><code class="language-r">cran_help_pages_links_wo_deps(pkg)
#&gt; [1] Package Source  Anchor  Target 
#&gt; &lt;0 rows&gt; (or 0-length row.names)
</code></pre>
<p>If there is some output then the link cannot be resolved correctly if the other package is not independently installed on the same machine.</p>
<h1 id="chp:reproducibility_1">Reproducibility</h1>
<p>If you wish to know what packages were available on CRAN on any given date you can use:</p>
<pre><code class="language-r">cs &lt;- cran_snapshot(as.Date(&quot;2020-01-31&quot;))
#&gt; Warning: There are 4 packages both archived and published
#&gt; This indicate manual CRAN intervention.
#&gt; Retrieving comments, this might take a bit.
#&gt; Caching results to be faster next call in this session.
nrow(cs)
#&gt; [1] 101420
</code></pre>
<p>This might be helpful to know what was available on old project and why some feature of a given package wasn’t used.
Maybe it wasn’t available on a given date!</p>
<h3 id="sec:local-versions">Local versions</h3>
<p>While working it might be good to update packages.
To decide if it is needed maybe you’d like to know when were packages last updated on the system?</p>
<pre><code class="language-r">cran_session()
#&gt; Warning in cran_archive(versions[, &quot;Package&quot;]): Omitting packages repo.data.
#&gt; Maybe they were not on CRAN?
#&gt; [1] &quot;2025-08-27 18:40:06 CEST&quot;
</code></pre>
<p>This uses the <code>sessionInfo()</code> output to find the date of last installation.
Under the hood it uses a function for an arbitrary packages and their versions:</p>
<pre><code class="language-r">versions &lt;- data.frame(Package = c(&quot;dplyr&quot;, &quot;Rcpp&quot;, &quot;rlang&quot;),
                       Version = c(&quot;1.1.4&quot;, &quot;0.8.9&quot;, NA))
cran_date(versions)
#&gt; [1] &quot;2023-11-17 17:50:03 CET&quot;
</code></pre>
<p>This is the first date were these packages were at the requested version number (or available).
Currently these packages can have a release with higher version numbers (this can be easily checked with <code>old.packages()</code>).</p>
<p>To answer the original question of this section we can use:</p>
<pre><code class="language-r">ip &lt;- cran_date(installed.packages())
#&gt; Warning in cran_archive(versions[, &quot;Package&quot;]): Omitting packages repo.data, AnnotationDbi, AnnotationFilter, AnnotationHub, BioCor, Biobase, BiocFileCache, BiocGenerics, BiocIO, BiocParallel, BiocPkgTools, BiocStyle, BiocVersion, Biostrings, DESeq2, DelayedArray, ExperimentHub, GO.db, GOSemSim, GSEABase, GenomeInfoDb, GenomeInfoDbData, GenomicAlignments, GenomicFeatures, GenomicRanges, HDF5Array, IRanges, KEGGREST, MatrixGenerics, ProtGenerics, Rhdf5lib, Rhtslib, Rsamtools, S4Arrays, S4Vectors, SingleCellExperiment, SparseArray, SummarizedExperiment, UCSC.utils, XVector, airway, alabaster.base, alabaster.matrix, alabaster.ranges, alabaster.sce, alabaster.schemas, alabaster.se, annotate, assorthead, biocViews, biomformat, cransays, ensembldb, fgsea, gypsum, h5mread, microshades, org.Hs.eg.db, phyloseq, preprocessCore, reactome.db, resios, rhdf5, rhdf5filters, rostemplate, rotemplate, rtracklayer, rutils, scRNAseq.
#&gt; Maybe they were not on CRAN?
ip
#&gt; [1] &quot;2025-08-29 16:00:06 CEST&quot;
</code></pre>
<h3 id="sec:risk-of-being-archived">Risk of being archived</h3>
<p>If you ever wonder which packages are at risk of being archived you can use <code>cran_doom()</code>:</p>
<pre><code class="language-r">cd &lt;- cran_doom(bioc = TRUE)
#&gt; Retrieving CRAN_db, this might take a bit.
#&gt; Caching results to be faster next call in this session.
#&gt; Retrieving bioc_available_release, this might take a bit.
#&gt; Caching results to be faster next call in this session.
cd[c(&quot;time_till_last&quot;, &quot;last_archived&quot;, &quot;npackages&quot;)]
#&gt; $time_till_last
#&gt; Time difference of 61 days
#&gt; 
#&gt; $last_archived
#&gt; [1] &quot;2025-10-31&quot;
#&gt; 
#&gt; $npackages
#&gt;  CRAN   all 
#&gt; 22625 26331
knitr::kable(head(cd$details))
</code></pre>
<table>
<thead>
<tr>
<th align="left">Package</th>
<th align="left">Deadline</th>
<th align="left">type</th>
<th align="left">repo</th>
<th align="right">n_affected</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">xportr</td>
<td align="left">2025-08-25</td>
<td align="left">direct</td>
<td align="left">CRAN</td>
<td align="right">6</td>
</tr>
<tr>
<td align="left">duckplyr</td>
<td align="left">2025-08-25</td>
<td align="left">direct</td>
<td align="left">CRAN</td>
<td align="right">2</td>
</tr>
<tr>
<td align="left">CopSens</td>
<td align="left">2025-09-01</td>
<td align="left">direct</td>
<td align="left">CRAN</td>
<td align="right">11</td>
</tr>
<tr>
<td align="left">SPARSEMODr</td>
<td align="left">2025-09-01</td>
<td align="left">direct</td>
<td align="left">CRAN</td>
<td align="right">11</td>
</tr>
<tr>
<td align="left">NIMAA</td>
<td align="left">2025-09-01</td>
<td align="left">direct</td>
<td align="left">CRAN</td>
<td align="right">8</td>
</tr>
<tr>
<td align="left">tidyrgee</td>
<td align="left">2025-09-01</td>
<td align="left">direct</td>
<td align="left">CRAN</td>
<td align="right">8</td>
</tr>
</tbody>
</table>
<p>There are website dedicated to track those and provide information about new version submissions to CRAN to fix those.
I participate on the <a href="https://www.cranhaven.org/dashboard-at-risk.html">cranhaven.org dashboard</a> (and project).</p>
<p>Note that if a package is archived it can be brought back to the repository.</p>
<h1 id="chp:reproducibility_2">Reproducibility</h1>
<p>For reproducibility here is the session info:</p>
<pre><code class="language-r">sessionInfo()
#&gt; R version 4.5.1 (2025-06-13)
#&gt; Platform: x86_64-pc-linux-gnu
#&gt; Running under: Ubuntu 24.04.3 LTS
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0
#&gt; 
#&gt; locale:
#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#&gt;  [3] LC_TIME=ca_ES.UTF-8        LC_COLLATE=C              
#&gt;  [5] LC_MONETARY=ca_ES.UTF-8    LC_MESSAGES=en_US.UTF-8   
#&gt;  [7] LC_PAPER=ca_ES.UTF-8       LC_NAME=C                 
#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#&gt; [11] LC_MEASUREMENT=ca_ES.UTF-8 LC_IDENTIFICATION=C       
#&gt; 
#&gt; time zone: Europe/Madrid
#&gt; tzcode source: system (glibc)
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] repo.data_0.0.0.9004
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] compiler_4.5.1    magrittr_2.0.3    cli_3.6.5         rversions_2.1.2  
#&gt;  [5] tools_4.5.1       igraph_2.1.4      rstudioapi_0.17.1 curl_7.0.0       
#&gt;  [9] xml2_1.4.0        knitr_1.50        xfun_0.53         lifecycle_1.0.4  
#&gt; [13] pkgconfig_2.0.3   rlang_1.1.6       evaluate_1.0.5
</code></pre>
</div>
</body>
</html>
